### DISCUSSION TITLE
Unity classes performance

### DISCUSSION THREAD

**Discussion by Deytooh** [naverz/zepeto-studio-global#18] [Q&A]
Hi ~~
So I was converting some C# code to typescript code...and I noticed some performance issues.

![image](https://user-images.githubusercontent.com/97385930/172571837-5f419933-1fe6-40d9-8da8-efaa4b5b4c4c.png)

```
    private Min(a: float, b: float): float { return a < b ? a : b; }


 Profiler.BeginSample("Custom Min");
        for (let i = 0; i < 2000000; i++) {
            this.Min(1,2);
        }
        Profiler.EndSample();

        Profiler.BeginSample("Mathf Min");
        for (let i = 0; i < 2000000; i++) {
            Mathf.Min(1,2);
        }
        Profiler.EndSample();
```

It is quite...very very very very different haha ! I guess if we use Unity classes, it somehow converts code through typescript / javascript / C# so it makes a lot of overload for it.

Is there any other class we can use instead of Mathf ? or do we need to build our own ?

**Comment by jhslee** [comment]
Hello @Deytooh 👋 

Hey nice catch!!! 🎉 Yes, it does seem like the unity Mathf library is much less performant than the custom script you've written. Please let me know if you notice other similar performance differences as well. I think as a general rule of thumb, its better to try and avoid unity classes as much as possible for now in favor of native javascript/typescript functions for anything that needs to happen over many consecutive frames until we can figure out whats going on here. 

For example, `Math.min(num1, num2)` might be a better alternative to `Mathf.Min(num1, num2)`
I'm attaching a link to the [Javascript Math class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math) for future reference. 

C# Mathf.Min
![image](https://user-images.githubusercontent.com/105688184/172960362-78f39f86-205a-4faa-80ea-7324ee3114db.png)

Typescript Mathf.min
![image](https://user-images.githubusercontent.com/105688184/172960733-1372f670-da4f-4048-a440-b2aa2f90f778.png)

Typescript Math.min
![image](https://user-images.githubusercontent.com/105688184/172958992-6ad28682-5e6f-4941-a237-5d4999f5f740.png)

**Reply by Deytooh** [reply]
Hello ~~ ;) Thank you for taking the time to answer me and do the tests :) 

I didnt even think about using the Math class from javascript indeed xD It will be useful ( and faster ) ! 

Actually every scripts that are originally from UnityEngine have all performances issues, for exemple I needed to call "GeometryUtility.TestPlanesAABB" a lot of times by frame, but it was making my game very laggy, so I wrote my custom TestPlanesAABB function,  custom Plane class ,custom Bounds, and a custom Vector3 class, and everything was muuuuuuuch faster ! 
I dont know why exactly but yeah I guess the Puerts engine is trying to find the Unity methods through a kind of heavy way ? 

So for doing pure calculations it's much better to write our own classes ( Vector3 etc... ) ( there is the UnityEngine code on Github so just need to convert to typescript :) )

**Reply by jhslee** [reply]
Hello @Deytooh!! 👋 
Yep, it looks like we're actually aware of these performance issues internally. There is indeed a fundamental difference in the performance between calling a method within a C# wrapper, and a library native to typescript. Because data is structured differently between C# and Typescript, there are processes set in place in order to overcome those differences that do negatively impact performance, such as the process of converting data types between each other, for example. 

We are actively looking for ways to optimize the "code flow" and improve performance, and we already do have a few ideas we're working on for a future update, but there is no timeline I can give at this time. Unfortunately, this means that until then you'll have to find ways, (like you already have 😊 ) to try and circumvent these for the time being.

I apologize for any inconvenience that this may cause, and hope for your understanding 🙏 

-John
